Edit 31.03.2022: Corrigendum: 

* [Corrigendum_Bioinformatics2021Weidner](https://github.com/sysbio-bioinf/BNStatic/blob/master/plotscripts_data/Corrigendum_Bioinformatics2021Weidner.pdf)
* [Corrigendum_Supplement](https://github.com/sysbio-bioinf/BNStatic/blob/master/plotscripts_data/Corrigendum_Supplement.pdf)

Edit 18.08.2021: Bugfix in the counting of network motifs.

All analyses and plots can be generated via the 'walkthrough.R' file in the folder 'plotscripts_data'.
First, the necessary libraries must be loaded. 

The variable 'pathtoscripts' has to be modified to point to the 'plotscripts_data' folder.
The set of Boolean networks to be analysed should be located in the 'Networks' folder in .txt format.

In the following, the available scripts and their results will be explained.

The script 'findthreshold_sugesttargets.R' can be used to generate the data (static and dynamic measures) for all networks.
It then determines the selection threshold comparing the selection of VBnDP (intersection) to the average of the three dynamic measures.
The nodes in a given network can then be classified and suggested intervention targets are returned. These are PM nodes, labelled as either hubs or non-hubs (a hub being a node with a connectivity z-score greater than or equal to 2.5). The nodes are sorted in order of decreasing mismatch.

### Data analysis

The script 'walkthrough.R' can be used to generate all plots shown in the paper and the supplement.


Calculate static and dynamic measures for all networks, save data

  This script calculates the static measures of Vertex Betweenness and Determinative Power as well as the 
  dynamic measures of Hamming distance, Attractor loss and Attractor gain for all networks located in the 'Networks' folder.
  These results have already been saved in the 'results' folder as RDS files and will be loaded 
  by the following scripts from there.
  
  
Calculate sensitivity & specificity for every threshold of selection

  The chosen static measures of VB, DP or their union/intersection are compared to the dynamic measures of Hamming distance,
  attractor loss and attractor gain given all possible selection thresholds T={1,...,100}.
  Sensitivity and specificity are calculated for all selections. 
  The average of these results across networks is used for the plots in Figures 1A+B.
  The results generated by this script are also already saved in the results folder 
  and will be loaded by the scripts for plotting.
  
  
Calculate max. MI along all simple paths to Hubs (Fig3D)

  Generates the data necessary for the plot in Figure 3D. 
  This data has also already been saved in the results folder and will
  be loaded from there for the following plot.
  This script calculates all simple paths leading from non-hubs to hubs and averages 
  the Mutual Information of the edges along these paths. 
  The maximal MI values for all start/end-node combinations are summed up and normalised by the number of hubs
  and the number of non-hubs of the given class present in the network.
  ==> PM nodes have paths with higher MI leading to hubs than NM or NS nodes.
  
  
### Plotting
  
Plot Figure 2A/S3
  The sensivity and specificity across selection thresholds for given static and dynamic measures are loaded.
  The resulting figure is obtained as the average of three sensitivity-specificity curves for the individual dynamic measures.
  => Find selection threshold as point where sensitivity==specificity, used for classifying nodes as non-selected (NS)
  

Plot Figure 2B/S4
  This plots shows the results from Fig1A split up across the individual dynamic measures of Hamming distance, 
  attractor loss and attractor gain.
  
  
  In the following, nodes are given two labels. They are either Hubs or Non-Hubs (by the chosen Hub definition).
  They also belong to one of the classes PM/NM/NS depending on whether they were selected by the static measure
  of VBnDP at the calculated threshold and on whether their position in the ranking of VBnDP is higher than their
  position in the ranking of connectivities.
  The rankings of all nodes belonging to the same class in a given network is then averaged and the results are shown
  across networks.
  In order to generate grouped plots showing both hub definitions simultaneously, 
  the previous script should be run once for both cases.
  
  The classification of nodes is also listed in the 'plotscripts_data/Node classification tables' folder.
  Hub nodes are colored red in their respective mismatch class.
  
  
Plot Figure 3A - Static impact
  A static impact of 1 indicates the node to have the highest score on the ranking of VBnDP, while 0 indicates the lowest.
  

Plot Figure 3B - Dynamic impact
  A dynamic impact of 1 indicates the node to have the highest score on the average of the three dynamic rankings, 
  while 0 indicates the lowest.
  
  
Plot Figure 3C - Connectivity
  The z-scores for all nodes belonging to a given class in a network is averaged, results are shown across networks.
  
  
Plot Figure 3D - Max. MI along paths to Hubs
  The generated data is loaded.
  The higher the values of the maximal MI, the more nodes will determine the state of their outputs.
  That is, a value of 1 would indicate that a signal is perfectly passed on from the start to the end node.
  A value of 0 would indicate that other regulatory influences would overrule the signal being passed on from the start node.
  

Bootstrap robustness test of threshold (FigS5)

  A seed containing 10 000 repeats of random selections of 35 networks is loaded.
  The sensitivity and specificity curves for the static measure of VBnDP (intersection) for these selections are 
  loaded and averaged.
  From this, the threshold at which sensitivity and specificity are equal is determined for all dynamic measures
  as well as the average of dynamic measures. The distribution of these thresholds is then plotted.
  ==> The threshold remains stable for the average of dynamic measures


Overlap between static and dynamic based selections (FigS6)
  Plots the overlap of genes with high dynamic impact with those selected by static measures, depending on the size of the network.

Frequency of network motif participation by class (FigS7)
  Plots the frequency with which genes in a given network participate in network motifs (C1-C4FFL, I1-I4FFL and Bifan), grouped by selected/non-selected or class.

Overlap with FVS (FigS8)
  Plots the overlap of nodes in the selected/non-selected category or a given class with the feedback vertex set (FVS) across networks.

Venn diagrams comparing gatekeepers to network motifs, canalyzers and the FVS (FigS9)
  The frequency with which genes in a network participate in coherent and incoherent feed-forward loops as well as bifan motifs is counted. Alternatively, these values can be loaded directly from the resultmatrix.RDS file.
  Two Venn diagrams are created. The first shows the overlap of nodes in the gatekeeper class with nodes participating in at least one network motif of any kind, nodes which canalyze at least one Boolean function in their network and nodes belonging to the feedback vertex set (FVS).
  The second diagram compares gatekeeper nodes to nodes participating at least once in any individual network motif.
  
  
  
  
